src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs

```diff
@@ -0,0 +1,824 @@
+            }
+        }
+
+        private static void ReadCallback(IAsyncResult transportResult)
+        {
+            if (!(transportResult.AsyncState is LazyAsyncResult))
+            {
+                if (GlobalLog.IsEnabled)
+                {
+                    GlobalLog.Assert("ReadCallback|State type is wrong, expected LazyAsyncResult.");
+                }
+
+                Debug.Fail("ReadCallback|State type is wrong, expected LazyAsyncResult.");
+            }
+
+            if (transportResult.CompletedSynchronously)
+            {
+                return;
+            }
+
+            LazyAsyncResult lazyResult = (LazyAsyncResult)transportResult.AsyncState;
+
+            // Async completion.
+            try
+            {
+                NegoState authState = (NegoState)lazyResult.AsyncObject;
+                byte[] message = authState._framer.EndReadMessage(transportResult);
+                authState.ProcessReceivedBlob(message, lazyResult);
+            }
+            catch (Exception e)
+            {
+                if (lazyResult.InternalPeekCompleted)
+                {
+                    // This will throw on a worker thread.
+                    throw;
+                }
+
+                lazyResult.InvokeCallback(e);
+            }
+        }
+
+        internal static bool IsError(SecurityStatusPal status)
+        {
+            return ((int)status >= (int)SecurityStatusPal.OutOfMemory);
+        }
+
+        private unsafe byte[] GetOutgoingBlob(byte[] incomingBlob, ref Exception e)
+        {
+            SecurityStatusPal statusCode;
+            byte[] message = _context.GetOutgoingBlob(incomingBlob, false, out statusCode);
+
+            if (IsError(statusCode))
+            {
+                e = CreateExceptionFromError(statusCode);
+                uint error = (uint)e.HResult;
+
+                message = new byte[8];  //sizeof(long)
+                for (int i = message.Length - 1; i >= 0; --i)
+                {
+                    message[i] = (byte)(error & 0xFF);
+                    error = (error >> 8);
+                }
+            }
+
+            if (message != null && message.Length == 0)
+            {
+                message = s_emptyMessage;
+            }
+
+            return message;
+        }
+
+        internal int EncryptData(byte[] buffer, int offset, int count, ref byte[] outBuffer)
+        {
+            CheckThrow(true);
+
+            // SSPI seems to ignore this sequence number.
+            ++_writeSequenceNumber;
+            return _context.Encrypt(buffer, offset, count, ref outBuffer, _writeSequenceNumber);
+        }
+
+        internal int DecryptData(byte[] buffer, int offset, int count, out int newOffset)
+        {
+            CheckThrow(true);
+
+            // SSPI seems to ignore this sequence number.
+            ++_readSequenceNumber;
+            return _context.Decrypt(buffer, offset, count, out newOffset, _readSequenceNumber);
+        }
+    }
+}
```

Comment:
I can't easily tell in the diff... was any of this logic changed when it was moved to this new file, or is it all the same as in the original version?